### PR TITLE
MAINT: Bump Cython version

### LINKED ISSUE #7987: Bump minimum Cython version to 0.25?
I believe the current minimum required version is 0.23.4:

https://github.com/scipy/scipy/blob/ab5ef39601b17b204f32c787644bb516350b1a76/pyproject.toml#L5

The current version is 0.27. My primary interest in bumping to 0.25 is that it fixes a complex division bug:

https://github.com/cython/cython/blob/master/CHANGES.rst#significant-bugs-fixed

which we have an ugly workaround for in `special`. (Implementing Smith's algorithm ourselves.)

### PR DESCRIPTION
**Pull request by ghost**
[ci-skip]

Closes #7987.

### CODE REVIEW

**Comment by person142** [comment]
One more version to bump here: https://github.com/scipy/scipy/blob/master/tools/cythonize.py#L58.

Could be others I don't know about.

File: tools/cythonize.py [hunk 1]
```diff
@@ -55,8 +56,25 @@ def process_pyx(fromfile, tofile, cwd):
     try:
         from Cython.Compiler.Version import version as cython_version
         from distutils.version import LooseVersion
-        if LooseVersion(cython_version) < LooseVersion('0.23.4'):
-            raise Exception('Building SciPy requires Cython >= 0.23.4')
+
+        # Try to find pyproject.toml
+        pyproject_toml = os.path.join(dirname(dirname(__file__)), "pyproject.toml")
```
Comment (person142) [inline tools/cythonize.py]: Ha yes that's a more elegant way to solve the problem.

Comment (xoviat) [reply tools/cythonize.py]: It works!

```
Building, see build.log...
Running scipy/linalg/_generate_pyx.py
Running scipy/special/_generate_pyx.py
Processing scipy/_lib/_ccallback_c.pyx
Processing scipy/_lib/messagestream.pyx
Processing scipy/cluster/_hierarchy.pyx
Traceback (most recent call last):
  File "/home/ubuntu/scipy/tools/cythonize.py", line 283, in <module>
    main()
  File "/home/ubuntu/scipy/tools/cythonize.py", line 279, in main
    find_process_files(root_dir)
  File "/home/ubuntu/scipy/tools/cythonize.py", line 268, in find_process_files
    for result in pool.imap(lambda args: process(*args), jobs):
  File "/opt/circleci/python/2.7.11/lib/python2.7/multiprocessing/pool.py", line 668, in next
    raise value
Exception: Building SciPy requires Cython >= 0.25
Processing scipy/cluster/_optimal_leaf_ordering.pyx
Processing scipy/cluster/_vq.pyx
Processing scipy/interpolate/_bspl.pyx
Processing scipy/interpolate/_ppoly.pyx
Processing scipy/interpolate/interpnd.pyx
Processing scipy/io/matlab/mio5_utils.pyx
Processing scipy/io/matlab/mio_utils.pyx
Processing scipy/io/matlab/streams.pyx
Processing scipy/linalg/_decomp_update.pyx.in
```

Comment (pv) [reply tools/cythonize.py]: I'd prefer `join(dirname(__file__), '..', '..', 'pyproject.toml')` over the iterated dirname.

Comment (xoviat) [reply tools/cythonize.py]: I think you have an extra `..` in there, but will do.

**Comment by person142** [comment]
Thanks @xoviat.

**Pull request merged**